extensions/azurecore/src/account-provider/utils/msalCachePlugin.ts

```diff
@@ -18,35 +19,54 @@ export class MsalCachePluginProvider {
 		Logger.verbose(`MsalCachePluginProvider: Using cache path ${_msalFilePath} and serviceName ${_serviceName}`);
 	}
 
+	private getLockfilePath(): string {
+		return this._msalFilePath + '.lock';
+	}
+
 	public getCachePlugin(): ICachePlugin {
+		const lockFilePath = this.getLockfilePath();
 		const beforeCacheAccess = async (cacheContext: TokenCacheContext): Promise<void> => {
-			let exists = true;
+			let isAccessible = true;
+			await this.WaitAndLock(lockFilePath);
 			try {
-				await fsPromises.access(this._msalFilePath, constants.R_OK | constants.W_OK);
-			} catch {
-				exists = false;
-			}
-			if (exists) {
 				try {
+					await fsPromises.access(this._msalFilePath, constants.R_OK | constants.W_OK);
+				} catch {
+					Logger.error(`MsalCachePlugin: Cache file is not accessible.`);
+					isAccessible = false;
+				}
+				if (isAccessible) {
 					const cache = await fsPromises.readFile(this._msalFilePath, { encoding: 'utf8' });
 					cacheContext.tokenCache.deserialize(cache);
 					Logger.verbose(`MsalCachePlugin: Token read from cache successfully.`);
-				} catch (e) {
-					Logger.error(`MsalCachePlugin: Failed to read from cache file. ${e}`);
+				}
+			} catch (e) {
+				Logger.error(`MsalCachePlugin: Failed to read from cache file: ${e}`);
+				// Handle JSON syntax error in cache file in case anything goes wrong.
+				if (e.stack.includes('SyntaxError')) {
+					// Clearing cache here will ensure account is marked stale so re-authentication can be triggered.
+					Logger.verbose(`MsalCachePlugin: Cache file contents will be cleeared.`);
+					await fsPromises.writeFile(this._msalFilePath, '', { encoding: 'utf8' });
+				} else {
 					throw e;
 				}
+			} finally {
+				lockFile.unlockSync(lockFilePath);
 			}
-		};
+		}
 
 		const afterCacheAccess = async (cacheContext: TokenCacheContext): Promise<void> => {
 			if (cacheContext.cacheHasChanged) {
+				await this.WaitAndLock(lockFilePath);
```

Comment:
Out of curiosity - have you checked what happens if the process crashes and leaves the lock file without being able to clean it up? Does this library handle that situation automatically or should we be adding special logic to handle that if it happens?

Reply:
`lockfile` seems to be taking care of stale lock files:
https://github.com/npm/lockfile/blob/master/test/basic.js#L153

Reply:
Hm, isn't that only if you set the stale option though?

Reply:
I'll give it a try and get back!

Reply:
Addressed this, we're now able to clear stale lock files with flag maintained.